core/src/main/java/kafka/server/ClientMetricsManager.java

```diff
@@ -16,31 +16,421 @@
+         since the last request, then the client instance will be re-evaluated. Validation of the
+         request will be done after the client instance is created. If client issues another get
+         telemetry request prior to push interval, then the client should get a throttle error but if
+         the subscription has changed since the last request then the client should get the updated
+         subscription immediately.
+        */
+        ClientMetricsInstance clientInstance = clientInstance(clientInstanceId, requestContext);
+
+        try {
+            // Validate the get request parameters for the client instance.
+            validateGetRequest(request, clientInstance, now);
+        } catch (ApiException exception) {
+            return request.getErrorResponse(0, exception);
+        }
+
+        clientInstance.lastKnownError(Errors.NONE);
+        return createGetSubscriptionResponse(clientInstanceId, clientInstance, throttleMs);
+    }
+
+    public PushTelemetryResponse processPushTelemetryRequest(PushTelemetryRequest request,
+        RequestContext requestContext, int throttleMs) {
+
+        Uuid clientInstanceId = request.data().clientInstanceId();
+        if (clientInstanceId == null || Uuid.RESERVED.contains(clientInstanceId)) {
+            String msg = String.format("Invalid request from the client [%s], invalid client instance id",
+                clientInstanceId);
+            return request.getErrorResponse(0, new InvalidRequestException(msg));
+        }
+
+        long now = time.milliseconds();
+        ClientMetricsInstance clientInstance = clientInstance(clientInstanceId, requestContext);
+
+        try {
+            // Validate the push request parameters for the client instance.
+            validatePushRequest(request, clientInstance, now);
+        } catch (ApiException exception) {
+            log.debug("Error validating push telemetry request from client [{}]", clientInstanceId, exception);
+            clientInstance.lastKnownError(Errors.forException(exception));
+            return request.getErrorResponse(0, exception);
+        } finally {
+            // Update the client instance with the latest push request parameters.
+            clientInstance.terminating(request.data().terminating());
+        }
+
+        // Push the metrics to the external client receiver plugin.
+        byte[] metrics = request.data().metrics();
+        if (metrics != null && metrics.length > 0) {
+            try {
+                ClientMetricsReceiverPlugin.instance().exportMetrics(requestContext, request);
+            } catch (Exception exception) {
+                clientInstance.lastKnownError(Errors.INVALID_RECORD);
+                return request.errorResponse(throttleMs, Errors.INVALID_RECORD);
+            }
+        }
+
+        clientInstance.lastKnownError(Errors.NONE);
+        return new PushTelemetryResponse(new PushTelemetryResponseData().setThrottleTimeMs(throttleMs));
```

Comment:
As discussed, we shouldn't set `throttleMs` here.

Reply:
As per the details mentioned earlier, that we shouldn't set `throttleMs` in case of errors from telemetry APIs else channel can be muted. But shouldn't we set `throttleMs` at all from telemetry APIs i.e. in case of successful response but broker actually needs to throttle the channel? I might be missing something here.


https://github.com/apache/kafka/pull/14699#discussion_r1391827581
https://github.com/apache/kafka/pull/14699#discussion_r1397714746

Reply:
Thanks @junrao for explaining this. I have removed throttleMs from PushTelemetry response. Also created below Jira to add changes in QuotaManager. Resolving the comment.

https://issues.apache.org/jira/browse/KAFKA-15863